### DISCUSSION TITLE
Is there any guidelines on how the "host" should leverage MCP?

### DISCUSSION THREAD

**Discussion by mii9000** [modelcontextprotocol/modelcontextprotocol#1243] [Q&A - Client implementation] [answered]
Basically trying to understand how Claude Desktop or any other host should use MCP. I did not find any clear guideline on it.

**Comment by sahuguet** [comment]
I see MCP as a language agnostic solution similar to what Agentic.so is doing to extend the capabilities of LLMs.

In their example (see https://github.com/transitive-bullshit/agentic?tab=readme-ov-file#intro), they have this magic function called `createAISDKTools`. I think what we are looking for is the equivalent for MCP.

```
...
import { generateText } from 'ai'
import { createAISDKTools } from '@agentic/ai-sdk'

import { WeatherClient } from '@agentic/stdlib'

const weather = new WeatherClient()
const result = await generateText({
  model: XXXXX,
  tools: createAISDKTools(weather),                              <==== WHERE SOME MAGIC HAPPENS
  toolChoice: 'required',
  prompt: 'What is the weather in San Francisco?'
})
```

**Comment by dsp-ant** [comment, marked as answer]
I think we should improve the documents with how to write hosts and what the guidelines are for hosts. Do you mind opening an issue in the docs repository about it?

**Comment by kinthaiofficial** [comment]
How the host should leverage MCP is a great question — the spec defines the protocol but leaves a lot of discretion to host implementations, and the decisions made at the host layer have big implications for security and capability.

A few guidelines that seem right based on what we've been building:

**Capability manifest before connection** — before an LLM host connects to an MCP server, it should fetch the server's capability manifest (analogous to A2A Agent Cards) to understand what tools are offered, what they cost, and what permissions they need. The host then decides whether to expose all tools or a subset to the current user/agent session.

**Progressive trust** — don't give all tools to all agents. The host should apply monotonic capability narrowing: each agent session gets only the tools it has demonstrated need for, not the full MCP server manifest. This is the host's job, not the server's.

**Budget mediation** — when an MCP tool call might consume significant resources (a web search, a database query, a file write), the host should enforce the agent's budget ceiling before the call. The server shouldn't be trusted to self-limit.

**Audit logging at the host** — the host is the right place to log all MCP tool calls with full context (which user, which agent, what was called, what was returned, how much it cost). The server sees individual calls; the host sees the full delegation context.

**Tool result sanitization** — before injecting tool results into the LLM context, the host should screen for prompt injection patterns. The server might return a web page that contains "ignore previous instructions..." — the host should catch this.

We've been thinking about these host-layer responsibilities in KinthAI's OpenClaw integration: https://blog.kinthai.ai/openclaw-multi-tenancy-why-vm-per-user-doesnt-scale — the multi-tenant isolation model.

Is there a reference implementation of a host that gets these right, or is this still mostly best-effort per-implementation?

**Comment by kinthaiofficial** [comment]
The host role in MCP is underspecified in the current spec — there's a lot of "here's what the protocol supports" but not much "here's how a host should make decisions." Some principles that seem to matter in practice:

**Tool selection is a host responsibility, not a model responsibility**
The host should present the model with only the tools relevant to the current task, not the full set of all connected MCP servers. This means the host needs some mechanism to classify tasks and filter tool sets accordingly. Otherwise every prompt gets overwhelmed with irrelevant tool descriptions.

**Permission escalation should go through the host, not the model**
When an agent wants to call a tool it doesn't currently have access to, that request should be mediated by the host (which can consult the user or check policy), not handled by the model trying to prompt-engineer its way around restrictions.

**The host owns the budget**
Cost attribution for tool calls should flow through the host. When agent A uses tool X, the host records the cost against agent A's budget. This requires the host to understand and track costs across all MCP tool invocations — the spec doesn't currently have a cost signaling mechanism.

**Consent and confirmation**
For consequential tool calls (write to database, send email, charge payment method), the host is the right layer for user confirmation — before forwarding the tool call to the MCP server, not after the model has "decided" to call it.

**Audit trail**
The host should maintain a tamper-evident log of all tool calls in a session: which tool, what args, what response, timestamp. This is critical for debugging and for proving agent behavior to auditors.

The A2A protocol and OpenClaw both have opinions on how agents should delegate capability — the MCP host spec could borrow from these patterns for the host-agent relationship.

**Comment by localden** [comment]
We've limited Discussions in this repository to meeting notes and are closing this thread. To continue the conversation, pick a channel from [Contributor Communication](https://modelcontextprotocol.io/community/communication). Thanks for posting.
